File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        min_p = max_p = nums[0]
        maximum_product = max_p

        for n in nums[1:]:
            temp = min_p
            min_p = min(n, min_p * n, max_p * n)
            max_p = max(n , n * temp, max_p * n)
            maximum_product = max(maximum_product, max_p)
        return maximum_product

File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int leftproduct = 1, rightproduct = 1, ans = nums[0];

        for(int i = 0; i < nums.length; i++)
        {
            leftproduct = leftproduct == 0 ? 1 : leftproduct;
            rightproduct = rightproduct == 0 ? 1 : rightproduct;

            leftproduct *= nums[i];
            rightproduct *= nums[nums.length-1 - i];

            ans = Math.max(ans, Math.max(leftproduct,rightproduct));
        }
        return ans;
    }
}